### ISSUE TITLE
Hello, I would like to ask, why is there such a big gap between the models trained by the focal loss of the cuda version and the pytorch version?

### ISSUE THREAD

**Issue by xc-chengdu** [open-mmlab/mmdetection#7907]

**Comment by jbwang1997** [comment]
Could you provide the results and implementations of the two models? Usually, a small fluctuation is normal in the deep learning model.

**Comment by xc-chengdu** [comment]
![image](https://user-images.githubusercontent.com/72692177/166848841-2bb8068d-9896-476f-a1b6-3c7b670ea119.png)
This is the non_cuda focal_loss
def py_sigmoid_focal_loss(pred,  # tensor(n,cls_out_channels)
                          target,  # (n,)
                          weight=None,  #(n,)
                          gamma=2.0,
                          alpha=0.25,
                          reduction='mean',
                          avg_factor=None):
    pred_sigmoid = pred.sigmoid()
    target = target.type_as(pred)
    pt = (1-pred_sigmoid) * target + pred_sigmoid * (1-target)
    focal_weight = (alpha * target + (1-alpha)*(1-target))*pt.pow(gamma)  # focal_loss权重
    loss = F.binary_cross_entropy_with_logits(pred, target, reduction='none') * focal_weight
    if weight is not None:
        weight = weight.view(-1, 1)
    loss = weight_reduce_loss(loss, weight, reduction, avg_factor)
    return loss


@LOSSES.register_module()
class FocalLoss(nn.Module):
    def __init__(self,
                 use_sigmoid=True,  
                 gamma=2.0,
                 alpha=0.25,
                 reduction='mean',  
                 loss_weight=1.0):  
        # 为True表示已经经过sigmoid计算，直接计算loss;否则，要先进行sigmoid，再计算loss
        super(FocalLoss, self).__init__()
        assert use_sigmoid is True, 'Only sigmoid focal loss supported now'
        self.use_sigmoid = use_sigmoid
        self.gamma = gamma
        self.alpha = alpha
        self.reduction = reduction
        self.loss_weight = loss_weight

    def forward(self,
                pred,
                target,
                weight=None,
                avg_factor=None,
                reduction_override=None):
        assert reduction_override in (None, 'none', 'mean', 'sum')
        reduction = (reduction_override if reduction_override else self.reduction)
        if self.use_sigmoid:
            num_classes = pred.size(1)
            target = F.one_hot(target, num_classes=num_classes + 1)
            target = target[:, :num_classes]
            loss_cls = self.loss_weight * py_sigmoid_focal_loss(
                pred,
                target,
                weight,
                gamma=self.gamma,
                alpha=self.alpha,
                reduction=reduction,
                avg_factor=avg_factor)
        else:
            raise NotImplementedError  
        return loss_cls

**Comment by xc-chengdu** [comment]
![image](https://user-images.githubusercontent.com/72692177/166849147-39318d49-0ecf-4a26-bb24-dd6e791ae2cc.png)

**Comment by xc-chengdu** [comment]
**_The non_cuda focal loss:_**
import torch.nn.functional as F
import torch.nn as nn
from ..registry import LOSSES
from .utils import weight_reduce_loss


def py_sigmoid_focal_loss(pred,  # tensor(n,cls_out_channels)
                          target,  # (n,)
                          weight=None,  #(n,)
                          gamma=2.0,
                          alpha=0.25,
                          reduction='mean',
                          avg_factor=None):
    pred_sigmoid = pred.sigmoid()
    target = target.type_as(pred)
    pt = (1-pred_sigmoid) * target + pred_sigmoid * (1-target)
    focal_weight = (alpha * target + (1-alpha)*(1-target))*pt.pow(gamma)  
    loss = F.binary_cross_entropy_with_logits(pred, target, reduction='none') * focal_weight
    if weight is not None:
        weight = weight.view(-1, 1)
    loss = weight_reduce_loss(loss, weight, reduction, avg_factor)
    return loss


@LOSSES.register_module()
class FocalLoss(nn.Module):
    def __init__(self,
                 use_sigmoid=True,
                 gamma=2.0,
                 alpha=0.25,
                 reduction='mean',
                 loss_weight=1.0):
        super(FocalLoss, self).__init__()
        assert use_sigmoid is True, 'Only sigmoid focal loss supported now'
        self.use_sigmoid = use_sigmoid
        self.gamma = gamma
        self.alpha = alpha
        self.reduction = reduction
        self.loss_weight = loss_weight

    def forward(self,
                pred,
                target,
                weight=None,
                avg_factor=None,
                reduction_override=None):
        assert reduction_override in (None, 'none', 'mean', 'sum')
        reduction = (reduction_override if reduction_override else self.reduction)
        if self.use_sigmoid:
            num_classes = pred.size(1)
            target = F.one_hot(target, num_classes=num_classes + 1)
            target = target[:, :num_classes]
            loss_cls = self.loss_weight * py_sigmoid_focal_loss(
                pred,
                target,
                weight,
                gamma=self.gamma,
                alpha=self.alpha,
                reduction=reduction,
                avg_factor=avg_factor)
        else:
            raise NotImplementedError
        return loss_cls

**Comment by xc-chengdu** [comment]
**The cuda version of focal loss, I directly call the official implementation of focal loss.**

**Comment by jbwang1997** [comment]
<img width="570" alt="image" src="https://user-images.githubusercontent.com/32032453/167761110-64e4c455-a558-40ed-afa9-cb6d4499ff51.png">
I have tested py_sigmoid_focal_loss on COCO dataset. The results look normal. I consider the fluctuation is caused by the dataset. The results may face a large fluctuation when the dataset is small.
